### ISSUE TITLE
Standard UCI for Castling is not being followed and its reflected on makeUci

### ISSUE THREAD

**Issue by andrescavallin** [niklasf/chessops#13]
Hello!!! 

This is the problem, Lichess Board API returns moves in UCI format. I compare those moves with the ones made on the DGT Board which are on SAN. When castling they don't match because lichess returns *e1g1* and **makeUci** returns *e1h1*

DGT SAN: O-O
lichess UCI: e1g1
chessops UCI: e1h1

The problem may be that **parseSan** calculates the **from** property correctly, but from my perspective, incorrectly the **to** property. Then the **makeUci** just takes the **to** and **from** properties and so, calculates a value for castling that does not follow the UCI standard.
UCI standard says that for castling the correct way is, for Normal Chess, *e1g1* . For 960 its may be *e1h1*
This problem happens on all four castlings.

I don't know if its better to solve this on **makeUci** or on **parseSan**, but I do think its worth fixing.

Script to reproduce:

```typescript
import { Chess, makeUci, NormalMove } from "chessops";
import { parseSan } from "chessops/san";
//Default board
const localBoard = Chess.default();
//Move Object
var moveObject: NormalMove | undefined;
//Ruy-Lopez kind of
var ruyLopez = ['e4','e5','Nf3','Nc6','Bb5','d6','O-O','Bg4','d3','Qd7','Qd2','O-O-O'];
//Conver to UCI
ruyLopez.forEach((sanMove) => {
    console.log(`SAN: ${sanMove}`);
    moveObject = <NormalMove>parseSan(localBoard, sanMove);
    if ( (moveObject) && localBoard.isLegal(moveObject))
    {
        var uciMove = makeUci(moveObject);
        console.log(`UCI: ${uciMove}`);
        console.log(`from: ${moveObject.from} to: ${moveObject.to}`);
        localBoard.play(moveObject);
    }
    console.log();
  });


```

Script Output:

```
SAN: e4
UCI: e2e4
from: 12 to: 28

SAN: e5
UCI: e7e5
from: 52 to: 36

SAN: Nf3
UCI: g1f3
from: 6 to: 21

SAN: Nc6
UCI: b8c6
from: 57 to: 42

SAN: Bb5
UCI: f1b5
from: 5 to: 33

SAN: d6
UCI: d7d6
from: 51 to: 43

SAN: O-O
UCI: e1h1
from: 4 to: 7

SAN: Bg4
UCI: c8g4
from: 58 to: 30

SAN: d3
UCI: d2d3
from: 11 to: 19

SAN: Qd7
UCI: d8d7
from: 59 to: 51

SAN: Qd2
UCI: d1d2
from: 3 to: 11

SAN: O-O-O
UCI: e8a8
from: 60 to: 56
```

**Comment by niklasf** [comment]
chessops always gives dests in UCI_Chess960 mode. It's a generalization that simplifies a lot of things, and it's by design.

However, both Lichess and chessops should always accept both kinds of notation (UCI and UCI_Chess960). So unless you have to compare moves from both sources (why?) everything should be fine. If you have to, you can use [`normalizeMove()`](https://niklasf.github.io/chessops/classes/_chess_.chess.html#normalizemove).

**Comment by andrescavallin** [comment]
Hello!! Thanks for the reply. I will try normalizeMove() then .
The reason I need to compare is that when you are making your opponent’s move on the DGT Board you can get it wrong and I need to alert the player so they can correct the move and keep the boards in sync.

**Comment by niklasf** [comment]
Ah, ok. Makes sense. Perhaps it would be more robust to compare positions instead of moves? I guess it depends on what you get from the DGT API.

**Comment by andrescavallin** [comment]
@niklasf comparing positions can help a lot since DGT Board sometimes after several illegal moves stops sending the SAN but still sends the position. Does chessops have a way to compare positions ? My dream will be to have a way to infer the move by comparing two positions, that way when DGT Board enters conflict, I can still detect the move. But is not urgent, there are other ways to solve the sync problem.

The normalizeMove() returned the same position, I did not noticed any change. I am not sure if the expected behavior of normalizeMove() is to fix the **to** property of the Move for standard chess.
If that is the expected behavior please let me know since that is exactly what I need.

This is what I did:

```typescript
    tmpMove = <NormalMove>parseSan(localBoard, sanMove);
    normalizedMoveObject = localBoard.normalizeMove(tmpMove);
```
If that is not the expected behaviour then I think I will handle this special condition on the DGT integration code (dgt/play.ts) and then I can improve on that if there is eventually a solution on chessops to return the UCI and not the UCI_Chess960

Eventually I will need to support UCI_Chess960 too

Full testing script

```typescript
import { Chess, makeUci, Move, NormalMove } from "chessops";
import { parseSan } from "chessops/san";
//Default board
const localBoard = Chess.default();
//Move Objects
var tmpMove: NormalMove;
var normalizedMoveObject: Move
//Ruy-Lopez kind of
var ruyLopez = ['e4','e5','Nf3','Nc6','Bb5','d6','O-O','Bg4','d3','Qd7','Qd2','O-O-O'];
var castlingSide: "a" | "h" | undefined;
//Conver to UCI
  ruyLopez.forEach((sanMove) => {
    console.log(`SAN: ${sanMove}`);
    tmpMove = <NormalMove>parseSan(localBoard, sanMove);
    normalizedMoveObject = localBoard.normalizeMove(tmpMove);
    if ( (normalizedMoveObject) && localBoard.isLegal(normalizedMoveObject))
    {
        castlingSide = localBoard.castlingSide(normalizedMoveObject);
        if (castlingSide)
            console.log('--Castling');
        var uciMove = makeUci(normalizedMoveObject);
        console.log(`UCI: ${uciMove}`);
        console.log(`from: ${tmpMove.from} to: ${normalizedMoveObject.to}`);
        localBoard.play(normalizedMoveObject);
    }
    console.log();
  });
```

```
SAN: e4
UCI: e2e4
from: 12 to: 28

SAN: e5
UCI: e7e5
from: 52 to: 36

SAN: Nf3
UCI: g1f3
from: 6 to: 21

SAN: Nc6
UCI: b8c6
from: 57 to: 42

SAN: Bb5
UCI: f1b5
from: 5 to: 33

SAN: d6
UCI: d7d6
from: 51 to: 43

SAN: O-O
--Castling
UCI: e1h1
from: 4 to: 7

SAN: Bg4
UCI: c8g4
from: 58 to: 30

SAN: d3
UCI: d2d3
from: 11 to: 19

SAN: Qd7
UCI: d8d7
from: 59 to: 51

SAN: Qd2
UCI: d1d2
from: 3 to: 11

SAN: O-O-O
--Castling
UCI: e8a8
from: 60 to: 56
```

**Comment by niklasf** [comment]
Regarding comparison of positions: Nothing built-in, but the tools are there. For example, the following untested code should return all the squares differing between two boards:

```typescript
function diff(a: Board, b: Board): SquareSet {
  return a.white.xor(b.white)
    .union(a.pawn.xor(b.pawn))
    .union(a.knight.xor(b.knight))
    .union(a.bishop.xor(b.bishop))
    .union(a.rook.xor(b.rook))
    .union(a.queen.xor(b.queen))
    .union(a.king.xor(b.king));
}
```

Regarding the issue from Discord: Oversight fixed in efbd3af379a26d9770a89f37d82fd4b5cd2b5fa9.

Regarding `normalizeMove()`: I should document these things. It's intended to produce the Chess960 compatible king-to-rook notation. No need to normalize moves coming from chessops, only from other sources. There is no built-in method for converting to two-step king moves, but it's not needed for comparing moves.

More comments about your code: `type CastlingSide = 'a' | 'h'` can be imported from `chessops/types`. `pos.isLegal(move)` is guaranteed for moves returned from `parseSan(pos, ...)`.
